radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -115,71 +192,155 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
 		}
+	}
 
+	private void processVerticesResponseForQCSync(Hash syncTo, SyncState syncState, GetVerticesResponse response) {
 		Vertex vertex = response.getVertices().get(0);
 		syncState.fetched.addFirst(vertex);
 		Hash nextVertexId = vertex.getQC().getProposed().getId();
 
 		if (vertices.containsKey(nextVertexId)) {
 			for (Vertex v: syncState.fetched) {
 				if (!addQC(v.getQC())) {
-					log.info("GET_VERTICES failed: {}", qc);
+					log.info("GET_VERTICES failed: {}", syncState.qc);
 					return;
 				}
 				try {
-					insertVertex(v);
+					insertVertexInternal(v);
 				} catch (VertexInsertionException e) {
 					log.info("GET_VERTICES failed: {}", e.getMessage());
 					return;
 				}
 			}
-			addQC(qc);
+			addQC(syncState.qc);
 		} else {
-			log.info("SYNC_VERTICES: Sending further GetVerticesRequest {}", nextVertexId);
+			log.info("SYNC_VERTICES: Sending further GetVerticesRequest for qc={} fetched={} root={}",
+				syncState.qc, syncState.fetched.size(), vertices.get(rootId.get()));
 			syncVerticesRPCSender.sendGetVerticesRequest(nextVertexId, syncState.author, 1, syncTo);
 		}
 	}
 
-	private void doSync(QuorumCertificate qc, ECPublicKey author) {
-		final Hash vertexId = qc.getProposed().getId();
-		if (syncing.containsKey(vertexId)) {
+	public void processGetVerticesResponse(GetVerticesResponse response) {
+		log.info("SYNC_VERTICES: Received GetVerticesResponse {}", response);
+
+		final Hash syncTo = (Hash) response.getOpaque();
+		SyncState syncState = syncing.get(syncTo);
+		if (syncState == null) {
 			return;
 		}
 
-		SyncState syncState = new SyncState(qc, author);
-		syncing.put(vertexId, syncState);
+		if (response.getVertices().isEmpty()) {
+			log.info("GET_VERTICES failed: {}", syncState.qc);
+			// failed
+			// TODO: retry
+			return;
+		}
 
-		log.info("SYNC_VERTICES: Sending initial GetVerticesRequest {}", vertexId);
+		switch (syncState.syncStage) {
+			case GET_COMMITTED_VERTICES:
+				processVerticesResponseForCommittedSync(syncTo, syncState, response);
+				break;
+			case GET_QC_VERTICES:
+				processVerticesResponseForQCSync(syncTo, syncState, response);
+				break;
+			default:
+				throw new IllegalStateException("Unknown sync stage: " + syncState.syncStage);
+		}
+	}
+
+	private void doQCSync(SyncState syncState) {
+		final Hash vertexId = syncState.getQC().getProposed().getId();
+		syncState.setSyncStage(SyncStage.GET_QC_VERTICES);
+		log.info("SYNC_VERTICES: QC: Sending initial GetVerticesRequest for qc={}", syncState.getQC());
 		syncVerticesRPCSender.sendGetVerticesRequest(vertexId, syncState.author, 1, vertexId);
 	}
 
+	private void doCommittedSync(SyncState syncState) {
+		final Hash vertexId = syncState.getQC().getProposed().getId();
+		syncState.setSyncStage(SyncStage.GET_COMMITTED_VERTICES);
+		log.info("SYNC_VERTICES: Committed: Sending initial GetVerticesRequest {}", syncState.getCommittedQC());
+		syncVerticesRPCSender.sendGetVerticesRequest(vertexId, syncState.author, 3, vertexId);
```

Comment:
What is the reason for the magic 3 here?

Reply:
want the three children of committedQC so that we can construct a valid committed root